proposals/0016-simd-application-fees.md

```diff
@@ -141,295 +125,367 @@ application is not CPI'd to. There were multiple possible approaches discussed
-an error `ReadLockDisabled`. If the transaction fails at this stage the payer
-will end up paying `base fees + prioritization fees + application fees`. If
-the transaction is successfully executed, then rebates are returned to the
-payer and the remaining fees are transferred to respective accounts. If the
-transaction fails to execute, the fees are transferred to respective accounts
-without any rebates.
-
-If payer has overpaid the application fees then after paying all application
-fees the remaining amount will be returned to the payer even if the
-transaction fails. In case of partial payment, the user will lose the
-partially paid amount. To avoid burning application fees or creating new
-accounts, we add a constraint that accounts on which application fees are paid
-must exist and write-locked. Suppose we pay application fees on an account
-that does not exist. In that case, the payer will end up paying
-`base fees + prioritization fees + application fees on existing accounts`, and
-the transaction will fail. This instruction cannot be CPI'ed into.
+This native program will implement all the features required to implement
+application fees.
+
+### PayApplicationFees Instruction
 
 It requires:
 
-Accounts :
+Accounts: None
 
-* None
 Argument: Maximum application fees to be paid in lamports as `u64`
 
-#### UpdateApplicationFees Instruction
+With this instruction, the fee payer accepts to pay application fees specifying
+the maximum amount. This instruction **MUST** be included in the transaction
+that interacts with dapps having application fees. This instruction is like an
+infallible transfer if the payer has enough funds i.e, even if the transaction
+fails, the payer will end up paying the required amount of application fees.
+This instruction raises the minimum required balance to `other fees` +
+`application fees`
+
+Special cases:
+
+- If the payer does not have enough balance, the transaction is not scheduled
+  and fails with an error `InsufficientFunds`. Consider this case the same as if
+  the payer does not have enough funds to pay the base fees.
+- If the payer did not include this instruction in the transaction that
+  write-locks accounts with application fees, then the transaction fails with an
+  error `ApplicationFeesNotPaid`. the payer will pay just `other fees`. The
+  failure is before the execution of the transaction.
+- If the payer underpays the application fees, it will be handled the same as if
+  application fees were not paid.
+- If the payer overpays the application fees, then the transaction will be
+  executed and the payer will be reimbursed the overpaid amount.
+
+This instruction cannot be CPI'ed into.
+
+### UpdateApplicationFees
+
+This mechanism will set the application fees for an account in the `Account`
+structure. The account must already exist and should be rent-free to change its
+application fees. Only the account authority must be allowed to update the
```

Comment:
it's *not* the account authority, programs owning the account are allowed to update, and can implement additional instructions with appropriate checks to let the authority configure it